skip tls verification if default transport is used with insecure option

komish opened the pull request:
Fixes #1553 

This PR addresses an inconsistency between the Crane library and the Crane CLI (which works correctly) with regards to interactions with image registries behind untrusted (read: self-signed) certificates.

This implementation internally tracks the Insecure/WithTransport options being set to allow `makeOptions` to decide whether or not a slightly modified default transport needs to be added to the remote options. If the right combination is not set, this does not add any default transport, relying on whatever default was to be used (same behavior as before).

This PR _does not_ change the Crane CLI because the `http.RoundTripper` that it builds out ultimately takes into account docker config values that a library caller should invoke manually. In doing so, a caller would be defining their own transport, and the logic here would not take effect. https://github.com/google/go-containerregistry/blob/main/cmd/crane/cmd/root.go#L88

Signed-off-by: Jose R. Gonzalez <jose@flutes.dev>

jonjohnsonjr commented on pkg/crane/options.go:
```diff
@@ -47,26 +51,43 @@ func makeOptions(opts ...Option) Options {
 
 // Insecure is an Option that allows image references to be fetched without TLS.
+// This will also allow for untrusted (e.g. self-signed) certificates in cases where
+// the default transport is used (i.e. when WithTransport is not defined).
```
```suggestion
// the default transport is used (i.e. when WithTransport is not used).
```

komish replied:
This suggestion should be incorporated in the latest force-push.

jonjohnsonjr reviewed (commented):
This LGTM overall. Do you mind adding a test to cover this? I'd probably be lazy and add a private `transport` to `Options` when you call `Insecure` and `WithTransport`, then just check that `InsecureSkipVerify` is set for that.

komish commented:
@jonjohnsonjr Sure thing. I went ahead and added tests to make sure we're setting the internal boolean values (that trigger the logic we want), and then a test to make sure we're adding the transport. As you suggested, I added a private transport value to facilitate the test, and rewired the logic to set that value. Made the test much simpler.

jonjohnsonjr commented on pkg/crane/options.go:
```diff
@@ -47,26 +52,44 @@ func makeOptions(opts ...Option) Options {
 func WithTransport(t http.RoundTripper) Option {
 	return func(o *Options) {
 		o.Remote = append(o.Remote, remote.WithTransport(t))
+		o.transportSet = true
```
What do you think about `o.transport = t` instead of the bool, then have `makeOptions` check for `opt.transport == nil` rather than having both the bool and the transport?

komish replied:
I'm thinking we'd still want to track the insecure bool, but otherwise that could work and removes a bit of a redundancy. I'll put it together and send it here momentarily (along with the Boilerplate check fix).

komish replied:
Implemented in https://github.com/google/go-containerregistry/commit/5112fe351dcc371dfe427daf1870978a4c08053e

jonjohnsonjr commented:
Thanks!
